<a href="https://colab.research.google.com/github/marsani/MachineLearning-2021/blob/main/Cleaning_Data01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Data set sumber untuk Data Cleaning
link di https://github.com/realpython/python-data-cleaning/tree/master/Datasets
ada tiga file csv, yaitu : (https://s.id/1orJD)
1. BL-Flickr-Images-Book.csv (Daftar Buku dari British Library),
2. university_towns.txt (Kota lokasi Sekolah Tinggi di US dan
3. olympics.csv (Partisipasi Semua Negara di Olimpiade Musim Dingin dan Musim Panas

In [69]:
#library pandas dan numpy
import pandas as pd
import numpy as np
#from functools import reduce

In [70]:
df=pd.read_csv("/content/drive/MyDrive/Dataset-2021/Cleaning/BL-Flickr-Images-Book.csv")
df.head()

Identifier             Edition Statement      Place of Publication  \
0         206                           NaN                    London   
1         216                           NaN  London; Virtue & Yorston   
2         218                           NaN                    London   
3         472                           NaN                    London   
4         480  A new edition, revised, etc.                    London   

  Date of Publication              Publisher  \
0         1879 [1878]       S. Tinsley & Co.   
1                1868           Virtue & Co.   
2                1869  Bradbury, Evans & Co.   
3                1851          James Darling   
4                1857   Wertheim & Macintosh   

                                               Title     Author  \
0                  Walter Forbes. [A novel.] By A. A      A. A.   
1  All for Greed. [A novel. The dedication signed...  A., A. A.   
2  Love the Avenger. By the author of “All for Gr...  A., A. A.   
3  Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
4  [The World in which I live, and my place in it...  A., E. S.   

                                   Contributors  Corporate Author  \
0                               FORBES, Walter.               NaN   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
2  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
3                   Appleyard, Ernest Silvanus.               NaN   
4                           BROOME, John Henry.               NaN   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   
2                     NaN          NaN       NaN   monographic   
3                     NaN          NaN       NaN   monographic   
4                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

In [23]:
#baca dataset
# DATASET DI gDRIVE https://s.id/1orXt
#pd.set_option('display.max_rows', None)
df=pd.read_csv('/content/drive/MyDrive/Dataset-2021/Cleaning/BL-Flickr-Images-Book.csv')
df.head() #baca 5 data awal
#print(df)

Identifier             Edition Statement      Place of Publication  \
0         206                           NaN                    London   
1         216                           NaN  London; Virtue & Yorston   
2         218                           NaN                    London   
3         472                           NaN                    London   
4         480  A new edition, revised, etc.                    London   

  Date of Publication              Publisher  \
0         1879 [1878]       S. Tinsley & Co.   
1                1868           Virtue & Co.   
2                1869  Bradbury, Evans & Co.   
3                1851          James Darling   
4                1857   Wertheim & Macintosh   

                                               Title     Author  \
0                  Walter Forbes. [A novel.] By A. A      A. A.   
1  All for Greed. [A novel. The dedication signed...  A., A. A.   
2  Love the Avenger. By the author of “All for Gr...  A., A. A.   
3  Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
4  [The World in which I live, and my place in it...  A., E. S.   

                                   Contributors  Corporate Author  \
0                               FORBES, Walter.               NaN   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
2  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
3                   Appleyard, Ernest Silvanus.               NaN   
4                           BROOME, John Henry.               NaN   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   
2                     NaN          NaN       NaN   monographic   
3                     NaN          NaN       NaN   monographic   
4                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

In [27]:
#deskripsi data
df.describe()

Corporate Author  Corporate Contributors  Engraver
count               0.0                     0.0       0.0
mean                NaN                     NaN       NaN
std                 NaN                     NaN       NaN
min                 NaN                     NaN       NaN
25%                 NaN                     NaN       NaN
50%                 NaN                     NaN       NaN
75%                 NaN                     NaN       NaN
max                 NaN                     NaN       NaN

** Kita definisikan daftar (list) nama dari semua kolom yang ingin kita drop. Kemudian jalankan perintah fungsi drop(), dengan melewatkan parameter inplace bernilai True dan parameter axis bernilai 1 **


In [71]:
#Drop Kolom
to_drop=['Edition Statement',
         'Corporate Author',
         'Former owner',
         'Engraver',
         'Contributors',
         'Issuance type',
         'Shelfmarks']
df.drop(to_drop,inplace=True,axis=1)

In [72]:
df.head()

Identifier      Place of Publication Date of Publication  \
0         206                    London         1879 [1878]   
1         216  London; Virtue & Yorston                1868   
2         218                    London                1869   
3         472                    London                1851   
4         480                    London                1857   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   

      Author  Corporate Contributors  \
0      A. A.                     NaN   
1  A., A. A.                     NaN   
2  A., A. A.                     NaN   
3  A., E. S.                     NaN   
4  A., E. S.                     NaN   

                                          Flickr URL  
0  http://www.flickr.com/photos/britishlibrary/ta...  
1  http://www.flickr.com/photos/britishlibrary/ta...  
2  http://www.flickr.com/photos/britishlibrary/ta...  
3  http://www.flickr.com/photos/britishlibrary/ta...  
4  http://www.flickr.com/photos/britishlibrary/ta...

Alternatif utk membuang kolom, dengan meneruskannya langsung ke parameter columns daripada memisahkan label-label yang mau dibuang:


In [ ]:
df.drop(columns=to_drop,inplace=True)

In [ ]:
df.head()

Identifier      Place of Publication Date of Publication  \
0         206                    London         1879 [1878]   
1         216  London; Virtue & Yorston                1868   
2         218                    London                1869   
3         472                    London                1851   
4         480                    London                1857   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   

      Author  Corporate Contributors  \
0      A. A.                     NaN   
1  A., A. A.                     NaN   
2  A., A. A.                     NaN   
3  A., E. S.                     NaN   
4  A., E. S.                     NaN   

                                          Flickr URL  
0  http://www.flickr.com/photos/britishlibrary/ta...  
1  http://www.flickr.com/photos/britishlibrary/ta...  
2  http://www.flickr.com/photos/britishlibrary/ta...  
3  http://www.flickr.com/photos/britishlibrary/ta...  
4  http://www.flickr.com/photos/britishlibrary/ta...

In [73]:
#mengubah indeks pada dataframe
df['Identifier'].is_unique

True

In [74]:
df=df.set_index('Identifier')
df.head()

Place of Publication Date of Publication  \
Identifier                                                 
206                           London         1879 [1878]   
216         London; Virtue & Yorston                1868   
218                           London                1869   
472                           London                1851   
480                           London                1857   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

            Corporate Contributors  \
Identifier                           
206                            NaN   
216                            NaN   
218                            NaN   
472                            NaN   
480                            NaN   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...

In [75]:
#206 adalah label pertama dari indeks.
df.loc[206]

Place of Publication                                                 London
Date of Publication                                             1879 [1878]
Publisher                                                  S. Tinsley & Co.
Title                                     Walter Forbes. [A novel.] By A. A
Author                                                                A. A.
Corporate Contributors                                                  NaN
Flickr URL                http://www.flickr.com/photos/britishlibrary/ta...
Name: 206, dtype: object

In [76]:
df.dtypes

Place of Publication       object
Date of Publication        object
Publisher                  object
Title                      object
Author                     object
Corporate Contributors    float64
Flickr URL                 object
dtype: object

In [77]:
df.dtypes.value_counts()

object     6
float64    1
dtype: int64

In [ ]:
df.set_index('Identifier',inplace=True)

In [ ]:
df.dtypes

Place of Publication       object
Date of Publication        object
Publisher                  object
Title                      object
Author                     object
Corporate Contributors    float64
Flickr URL                 object
dtype: object

In [ ]:
#df.get_dtype_counts()
df.dtypes.value_counts()

object     6
float64    1
dtype: int64

In [52]:
df.loc[1905:,'Publisher'].head(10)

Identifier
1905                  NaN
1929                  NaN
2836                  NaN
2854       E. Moxon & Co.
2956                  NaN
2957                  NaN
3017                  NaN
3131            W. Abbatt
4598           The Author
4884    J. Hatchard & Son
Name: Publisher, dtype: object

In [81]:
#Merapihkan Fields dalam  Data
#dibersihkan Date of Publication dan Place of Publication
df.loc[206:,'Date of Publication'].head(10)

Identifier
206     1879 [1878]
216            1868
218            1869
472            1851
480            1857
481            1875
519            1872
667             NaN
874            1676
1143           1679
Name: Date of Publication, dtype: object

Buku tertentu hanya memiliki satu tanggal publikasi
- Hilangkan tanggal lain dalam kurung siku, 1879[1878]
- Konversi rentang tanggal ke "start date", 1860-63; 1839, 38-54
- Hilangkan tanggal yang tidak jelas dan gantikan dengan NaN NumPy, 
  [1879?] -> NaN
- Konversi string nan ke nilai NaN NumPy

In [82]:
regex=r'^(\d{4})' #reguler expreesion
regex

'^(\\d{4})'

- Perintah \d mewakili sebarang digit dan {4} mengulangi aturan (rule) sebanyak empat kali. 
- Kararakter ^ sesuai dengan awal string, dan tanda dalam kurung () menunjukkan 
capturing group yang memberikan sinyal ke Pandas bahwa akan dilakukan ekstraksi 
bagian Regex tersebut

In [85]:
extr=df['Date of Publication'].str.extract(r'^(\d{4})',expand=False)
extr.head()

Identifier
206    1879
216    1868
218    1869
472    1851
480    1857
Name: Date of Publication, dtype: object

In [84]:
df.loc[206:,'Date of Publication'].head(10)

Identifier
206     1879 [1878]
216            1868
218            1869
472            1851
480            1857
481            1875
519            1872
667             NaN
874            1676
1143           1679
Name: Date of Publication, dtype: object

In [86]:
df['Date of Publication']=pd.to_numeric(extr)
df['Date of Publication'].dtype

dtype('float64')

In [87]:
df.loc[206:,'Date of Publication'].head(10)

Identifier
206     1879.0
216     1868.0
218     1869.0
472     1851.0
480     1857.0
481     1875.0
519     1872.0
667        NaN
874     1676.0
1143    1679.0
Name: Date of Publication, dtype: float64

In [88]:
df.to_csv('/content/drive/MyDrive/Dataset-2021/Cleaning/data2.csv')
df1=pd.read_csv('/content/drive/MyDrive/Dataset-2021/Cleaning/data2.csv')
df1.head() #baca 5 data awal

Identifier      Place of Publication  Date of Publication  \
0         206                    London               1879.0   
1         216  London; Virtue & Yorston               1868.0   
2         218                    London               1869.0   
3         472                    London               1851.0   
4         480                    London               1857.0   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   

      Author  Corporate Contributors  \
0      A. A.                     NaN   
1  A., A. A.                     NaN   
2  A., A. A.                     NaN   
3  A., E. S.                     NaN   
4  A., E. S.                     NaN   

                                          Flickr URL  
0  http://www.flickr.com/photos/britishlibrary/ta...  
1  http://www.flickr.com/photos/britishlibrary/ta...  
2  http://www.flickr.com/photos/britishlibrary/ta...  
3  http://www.flickr.com/photos/britishlibrary/ta...  
4  http://www.flickr.com/photos/britishlibrary/ta...

In [ ]:
extr

- kolom tsb masih memiliki object dtype, namun dapat kita konversi numeriknya dengan perintah pd.to_numeric

In [ ]:
df['Date of Publication']=pd.to_numeric(extr)
df['Date of Publication'].dtype

dtype('float64')

- Ini menghasilkan sekitar 1/10 nilai yang hilang, cost yang cukup kecil 
dampaknya untuk saat ini karena dapat melakukan perhitungan pada nilai valid yang tersisa

In [ ]:
df['Date of Publication'].isnull().sum()
#/len(df)

971

In [ ]:
len(df)

8287

In [ ]:
df['Date of Publication'].isnull().sum()/len(df)

0.11717147339205986

- Field place of publication masih ada informasi yang tidak penting. Jika 
dilihat lebih teliti, kasus ini untuk beberapa baris yang place of 
publication -nya di “London” dan “Oxford”.


In [ ]:
df['Place of Publication'].head(10)

Identifier
206                                  London
216                London; Virtue & Yorston
218                                  London
472                                  London
480                                  London
481                                  London
519                                  London
667     pp. 40. G. Bryan & Co: Oxford, 1898
874                                 London]
1143                                 London
Name: Place of Publication, dtype: object

In [ ]:
df.loc[4157862]

Place of Publication                                    Newcastle-upon-Tyne
Date of Publication                                                    1867
Publisher                                                        T. Fordyce
Title                     Local Records; or, Historical Register of rema...
Author                        FORDYCE, T. - Printer, of Newcastle-upon-Tyne
Corporate Contributors                                                  NaN
Flickr URL                http://www.flickr.com/photos/britishlibrary/ta...
Name: 4157862, dtype: object

- Gunakan str.contains() untuk mendapatkan Boolean mask

In [ ]:
pub=df['Place of Publication']
london=pub.str.contains('London')
london[:5]

Identifier
206    True
216    True
218    True
472    True
480    True
Name: Place of Publication, dtype: bool

In [ ]:
pub=df['Place of Publication']
london=pub.str.contains('London')
london

In [ ]:
oxford=pub.str.contains('Oxford')
#np.where(london,'London',np.where(exford,'Oxford'))
df['Place of Publication']= np.where(london, 'London',np.where(oxford,'Oxford',pub.str.replace('-',' ')))


In [ ]:
df['Place of Publication'].head()

Identifier
206    London
216    London
218    London
472    London
480    London
Name: Place of Publication, dtype: object

In [ ]:
df.head()

Place of Publication  ...                                         Flickr URL
Identifier                       ...                                                   
206                      London  ...  http://www.flickr.com/photos/britishlibrary/ta...
216                      London  ...  http://www.flickr.com/photos/britishlibrary/ta...
218                      London  ...  http://www.flickr.com/photos/britishlibrary/ta...
472                      London  ...  http://www.flickr.com/photos/britishlibrary/ta...
480                      London  ...  http://www.flickr.com/photos/britishlibrary/ta...

[5 rows x 7 columns]

In [ ]:
import pandas as pd
#df.to_csv('/content/drive/MyDrive/Dataset-2021/Cleaning/data.csv')
df1=pd.read_csv('/content/drive/MyDrive/Dataset-2021/Cleaning/data.csv')
df1.head() #baca 5 data awal

Identifier  ...                                         Flickr URL
0         206  ...  http://www.flickr.com/photos/britishlibrary/ta...
1         216  ...  http://www.flickr.com/photos/britishlibrary/ta...
2         218  ...  http://www.flickr.com/photos/britishlibrary/ta...
3         472  ...  http://www.flickr.com/photos/britishlibrary/ta...
4         480  ...  http://www.flickr.com/photos/britishlibrary/ta...

[5 rows x 8 columns]

In [ ]:
df1.dtypes

Identifier                  int64
Place of Publication       object
Date of Publication       float64
Publisher                  object
Title                      object
Author                     object
Corporate Contributors    float64
Flickr URL                 object
dtype: object

In [ ]:
df1['Date of Publication'].isnull()

0       False
1       False
2       False
3       False
4       False
        ...  
8282    False
8283    False
8284     True
8285    False
8286    False
Name: Date of Publication, Length: 8287, dtype: bool